Use the correct sequence for shard replacement

davisp opened the pull request:
The call to unpack the since sequence for _changes incorrectly filtered
the value for any shard that didn't exist. This means that any shard
that was moved ended up requiring a complete rescan which is a bug. This
just sets the update sequence to the range of the shard being replaced
rather than zero.

BugzID: 44287

rnewson commented:
Agreed this does what you said, but how can we just switch like this? Don't we need to find the previous checkpoint between the shards we're failing over between?

mikewallace1979 commented:
It looks like the assumption was that the replacement shard would be the shard copy which had been moved, and therefore the seq value from the previous since sequence could be safely used on the replacement shard without missing updates.

There are two potential issues:

 1. I can't see anything that would guarantee the rsync'd copy becomes the replacement shard in the since sequence - it looks like `Replacements` is a list of the other copies of that range, one of which is the rsync'd copy (and therefore with an (mostly) identical seq tree) and the others being regular shard copies with completely different seq trees.

 2. Even if we are swapping in the correct replacement shard it is still possible they will have divergent seq trees because the shard move process involves both the source and target shard being in the shard map at the same time. If any updates are received during that time then the ordering could be different on the source and target shard and therefore when the source is finally removed we could end up missing updates when we use the max_seq from the source shard for the replacement shard.

theburge commented:
Just to add to Mike's point 2, in case it's relevant:

We often find the leftovers from screwy shard moves, where we have N={4,5,6} and then have to knock one or more shards out of a given range. In that situation, those additional copies might have been in the shard map for months or more so they could be quite divergent...

We're really not very good at cleaning that stuff up in a timely way and clou doesn't guarantee to end the day on N=3.

davisp commented:
This bit of code is quite gnarly but there's a lot happening that's not necessarily apparent from a local inspection.

For concerns raised by @mikewallace1979 and @theburge, the relevant code to make those guarantees is in fabric_rpc2.erl here:

https://github.com/cloudant/fabric/blob/master/src/fabric_rpc2.erl#L578-L606

The way that we make the guarantee that we're only restarting from the rsync-ed copy is that we rely on the UUID match along with the epochs listing. Epochs list who started owning a file before the first write. This means that when its first opened on the target node we can look at who owns the file for particular updates sequences and rewind as necessary.

For Bob's point we only have to look at checkpoints when we're moving between non-rsync'ed copies which means that the UUIDs are different which prevents from this fastforward. Looking at fabric_rpc2 there's also apparently code already for this which I totally forgot we added. So that's fun.

Incidentally, this particular bit of code is incorrect in some cases because we have a terrible time generating update sequences that have any sense of sanity. We never managed to fix the screwed up difference between sequence values of Seq (as a number), {Seq, UUID} and {Seq, UUID, NodeName}. So I need to go back and add guards and substitutions to get those fiddly bits correct. Updated PR coming shortly.

davisp commented:
The change was to add the case statement patterns to munge the possible formats for the update sequence values here:

https://github.com/cloudant/fabric/commit/daa1b81082ee45d4d54ba5aa9e1852425e438767#diff-2af86e058b4e7a4a99a7c5a12da6debdR438

mikewallace1979 commented:
@davisp Thanks for the explanation - makes sense and LGTM (unsure how much that's worth coming from me but there it is).

theburge commented:
Yep, thanks, this change makes sense to me too now.
